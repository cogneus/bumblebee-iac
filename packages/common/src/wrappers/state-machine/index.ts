export * from './state-machine.wrapper'
export * from './types'
