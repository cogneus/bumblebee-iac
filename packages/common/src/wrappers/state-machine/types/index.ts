export * from './state-machine-dependencies.interface'
export * from './state-machine-handler.type'
export * from './state-machine-input.interface'
export * from './state-machine-payload.interface'
export * from './state-machine-wrapper.type'
export * from './state-machine-wrapper-params.interface'
