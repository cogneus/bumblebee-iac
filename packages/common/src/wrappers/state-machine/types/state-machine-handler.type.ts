import { StateMachineDependencies } from './state-machine-dependencies.interface'
import { StateMachinePayload } from './state-machine-payload.interface'

export type StateMachineHandler<TInput extends StateMachinePayload, TResult> = (
  args: StateMachineDependencies<TInput>
) => Promise<TResult>
