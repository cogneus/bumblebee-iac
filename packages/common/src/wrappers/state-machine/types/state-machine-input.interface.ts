import { StateMachinePayload } from './state-machine-payload.interface'

export interface StateMachineInput<T extends StateMachinePayload> {
  payload: T
}
