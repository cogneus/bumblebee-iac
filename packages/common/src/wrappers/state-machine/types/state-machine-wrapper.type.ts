import { Handler } from 'aws-lambda'
import { StateMachinePayload } from './state-machine-payload.interface'
import { StateMachineInput } from './state-machine-input.interface'
import { StateMachineWrapperParams } from './state-machine-wrapper-params.interface'

export type StateMachineWrapper = <TInput extends StateMachinePayload, TResult>(
  params: StateMachineWrapperParams<TInput, TResult>
) => Handler<StateMachineInput<TInput>, TResult>
