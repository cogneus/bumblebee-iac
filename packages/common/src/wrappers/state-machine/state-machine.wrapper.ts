import 'source-map-support/register'
import { ServiceContext } from '../../types'
import { createWrapperContext } from '../utils'
import { StateMachineWrapper } from './types'
import { destroyClients } from '../../clients'

export const stateMachineWrapper: StateMachineWrapper =
  ({ handler, getClientOptions, ...contextParams }) =>
  async (event, context: ServiceContext) => {
    const { payload } = event
    const serviceContext = createWrapperContext({
      ...contextParams,
      context,
      logInfo: {
        payload,
      },
      clientOptions: getClientOptions(payload),
    })
    const { log } = serviceContext
    try {
      log.debug('Process input', event)
      const result = await handler({
        payload,
        serviceContext,
      })
      log.debug('Process complete', result as any)
      return result
    } catch (ex) {
      log.warn(`Process failed, ${ex.message}`)
      log.error(ex)
      throw ex
    } finally {
      destroyClients(serviceContext.clients)
    }
  }
