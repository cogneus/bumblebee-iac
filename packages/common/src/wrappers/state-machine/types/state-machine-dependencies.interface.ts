import { ServiceContext } from '../../../types'
import { StateMachinePayload } from './state-machine-payload.interface'

export interface StateMachineDependencies<T extends StateMachinePayload> {
  payload: T
  serviceContext: ServiceContext
}
