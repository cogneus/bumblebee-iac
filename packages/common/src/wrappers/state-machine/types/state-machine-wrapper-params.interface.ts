import { ClientOptions } from '../../../clients'
import { StateMachinePayload } from './state-machine-payload.interface'
import { StateMachineHandler } from './state-machine-handler.type'
import { WrapperParams } from '../../types'

export interface StateMachineWrapperParams<TInput extends StateMachinePayload, TResult>
  extends WrapperParams<StateMachineHandler<TInput, TResult>, (payload: TInput) => ClientOptions> {}
