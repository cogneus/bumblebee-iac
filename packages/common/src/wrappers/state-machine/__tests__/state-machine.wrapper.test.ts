import { ClientOptions } from '../../../clients'
import { stateMachineWrapper } from '../state-machine.wrapper'
import { StateMachineInput, StateMachinePayload } from '../types'

const baseInput: StateMachineInput<StateMachinePayload> = {
  payload: {
    test: 'test',
  },
}

const events = [
  {
    input: baseInput,
    response: {
      result: 'success',
    },
  },
  {
    input: baseInput,
    response: {
      result: 'success',
    },
  },
  {
    input: baseInput,
    error: {
      message: 'failure',
    },
  },
]

const clientOptions: ClientOptions = {
  db: {
    maxAttempts: 3,
    region: 'eu-west-1',
    requestTimeout: 3000,
    tables: {
      test: {
        name: 'test-table-name',
        indices: {
          testIndex: { name: 'test-index-name' },
        },
        itemLimits: { defaultLimit: 20, maxLimit: 100 },
      },
    },
  },
}

describe('step function wrapper tests', () => {
  describe('should handle response, returns', () => {
    it.each(events)('results matching snapshot', async ({ input, response, error }) => {
      jest.resetAllMocks()
      const wrapper = stateMachineWrapper({
        handler: async ({ payload }) => {
          if (error) {
            throw error
          }
          return {
            ...payload,
            ...response,
          }
        },
        region: 'eu-west-1',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        env: 'dev',
        getClientOptions: () => clientOptions,
      })
      const cb: any = null
      let result: any = null
      let ex: any = null
      try {
        result = await wrapper(input, {} as any, cb)
      } catch (e) {
        ex = e
      }
      if (ex && error) {
        expect(ex.message).toMatch(error.message)
      } else {
        expect(result).toMatchSnapshot()
      }
    })
  })
})
