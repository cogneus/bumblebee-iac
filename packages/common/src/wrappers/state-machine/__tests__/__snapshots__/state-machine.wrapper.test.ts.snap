// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step function wrapper tests should handle response, returns results matching snapshot 1`] = `
{
  "result": "success",
  "test": "test",
}
`;

exports[`step function wrapper tests should handle response, returns results matching snapshot 2`] = `
{
  "result": "success",
  "test": "test",
}
`;
